diff --git a/crypto/openssl_util.cc b/crypto/openssl_util.cc
old mode 100644
new mode 100755
index 34af810..549b638
--- a/crypto/openssl_util.cc
+++ b/crypto/openssl_util.cc
@@ -58,8 +58,8 @@ class OpenSSLInitSingleton {
 #if defined(OS_ANDROID) && defined(ARCH_CPU_ARMEL)
     const bool has_neon =
         (android_getCpuFeatures() & ANDROID_CPU_ARM_FEATURE_NEON) != 0;
-    if (has_neon)
-      CRYPTO_set_NEON_capable(1);
+//    if (has_neon)
+//      CRYPTO_set_NEON_capable(1);
 #endif
   }
