diff --git a/roots.cpp b/roots.cpp
old mode 100644
new mode 100755
index d69245f..a924292
--- a/roots.cpp
+++ b/roots.cpp
@@ -34,7 +34,7 @@ extern "C" {
 #include "make_ext4fs.h"
 extern "C" {
 #include "wipe.h"
-#include "cryptfs.h"
+//#include "cryptfs.h"
 }
 
 static struct fstab *fstab = NULL;
@@ -244,7 +244,7 @@ int format_volume(const char* volume) {
         if (v->length != 0) {
             length = v->length;
         } else if (v->key_loc != NULL && strcmp(v->key_loc, "footer") == 0) {
-            length = -CRYPT_FOOTER_OFFSET;
+//            length = -CRYPT_FOOTER_OFFSET;
         }
         int result;
         if (strcmp(v->fs_type, "ext4") == 0) {
diff --git a/updater/Android.mk b/updater/Android.mk
old mode 100644
new mode 100755
index cc22061..a8e9fd2
--- a/updater/Android.mk
+++ b/updater/Android.mk
@@ -18,7 +18,7 @@ include $(CLEAR_VARS)
 LOCAL_MODULE_TAGS := eng
 
 LOCAL_SRC_FILES := $(updater_src_files)
-
+TARGET_USERIMAGES_USE_EXT4 := true
 ifeq ($(TARGET_USERIMAGES_USE_EXT4), true)
 LOCAL_CFLAGS += -DUSE_EXT4
 LOCAL_CFLAGS += -Wno-unused-parameter
diff --git a/updater/install.c b/updater/install.c
old mode 100644
new mode 100755
index 20ee431..33bd929
--- a/updater/install.c
+++ b/updater/install.c
@@ -50,10 +50,10 @@
 #include "install.h"
 #include "tune2fs.h"
 
-#ifdef USE_EXT4
+//#ifdef USE_EXT4
 #include "make_ext4fs.h"
 #include "wipe.h"
-#endif
+//#endif
 
 void uiPrint(State* state, char* buffer) {
     char* line = strtok(buffer, "\n");
@@ -1471,7 +1471,7 @@ Value* WipeBlockDeviceFn(const char* name, State* state, int argc, Expr* argv[])
 
     size_t len = strtoull(len_str, NULL, 0);
     int fd = open(filename, O_WRONLY, 0644);
-    int success = wipe_block_device(fd, len);
+    int success = wipe_block_device(fd, len); 
 
     free(filename);
     free(len_str);
