project bootable/recovery/
diff --git a/Android.mk b/Android.mk
old mode 100644
new mode 100755
index b159468..1901625
--- a/Android.mk
+++ b/Android.mk
@@ -101,7 +101,7 @@ ifeq ($(TARGET_HAVE_OEMLOCK), true)
     LOCAL_STATIC_LIBRARIES += liboemlock
 endif
 
-ifeq ($(TARGET_USERIMAGES_USE_EXT4), true)
+ifeq ($(TARGET_USERIMAGES_USE_EXT4), )
     LOCAL_CFLAGS += -DUSE_EXT4
     LOCAL_C_INCLUDES += system/extras/ext4_utils system/vold
     LOCAL_STATIC_LIBRARIES += libext4_utils_static libz
@@ -192,7 +192,7 @@ LOCAL_SRC_FILES := \
     roots.cpp
 LOCAL_CFLAGS += -DMINIVOLD
 LOCAL_CFLAGS += -Wno-unused-parameter
-ifeq ($(TARGET_USERIMAGES_USE_EXT4), true)
+ifeq ($(TARGET_USERIMAGES_USE_EXT4), )
     LOCAL_CFLAGS += -DUSE_EXT4
     LOCAL_C_INCLUDES += system/extras/ext4_utils
     LOCAL_STATIC_LIBRARIES += libext4_utils_static libz
diff --git a/updater/Android.mk b/updater/Android.mk
old mode 100644
new mode 100755
index f87cea5..9e78666
--- a/updater/Android.mk
+++ b/updater/Android.mk
@@ -19,7 +19,7 @@ LOCAL_MODULE_TAGS := eng
 
 LOCAL_SRC_FILES := $(updater_src_files)
 
-ifeq ($(TARGET_USERIMAGES_USE_EXT4), true)
+ifeq ($(TARGET_USERIMAGES_USE_EXT4), )
 LOCAL_CFLAGS += -DUSE_EXT4
 LOCAL_CFLAGS += -Wno-unused-parameter
 LOCAL_C_INCLUDES += system/extras/ext4_utils
