diff --git a/core/config.mk b/core/config.mk
old mode 100644
new mode 100755
index a7102ce..ba462dc
--- a/core/config.mk
+++ b/core/config.mk
@@ -636,7 +636,7 @@ $(warning COMMON_GLOBAL_C(PP)FLAGS changed)
 $(info *** Device configurations are no longer allowed to change the global flags.)
 $(info *** COMMON_GLOBAL_CFLAGS: $(COMMON_GLOBAL_CFLAGS))
 $(info *** COMMON_GLOBAL_CPPFLAGS: $(COMMON_GLOBAL_CPPFLAGS))
-$(error bailing...)
+# $(error bailing...)
 endif
 
 # These can be changed to modify both host and device modules.
