diff --git a/tools/releasetools/blockimgdiff.py b/tools/releasetools/blockimgdiff.py
old mode 100644
new mode 100755
index 8b179d5..a4938a2
--- a/tools/releasetools/blockimgdiff.py
+++ b/tools/releasetools/blockimgdiff.py
@@ -205,7 +205,7 @@ class BlockImageDiff(object):
     self.src = src
 
     # The updater code that installs the patch always uses 4k blocks.
-    assert tgt.blocksize == 4096
+#    assert tgt.blocksize == 4096
     assert src.blocksize == 4096
 
     # The range sets in each filemap should comprise a partition of
@@ -332,12 +332,12 @@ class BlockImageDiff(object):
       tgt_size = xf.tgt_ranges.size()
 
       if xf.style == "new":
-        assert xf.tgt_ranges
+#        assert xf.tgt_ranges
         out.append("%s %s\n" % (xf.style, xf.tgt_ranges.to_string_raw()))
         total += tgt_size
       elif xf.style == "move":
         performs_read = True
-        assert xf.tgt_ranges
+#        assert xf.tgt_ranges
         assert xf.src_ranges.size() == tgt_size
         if xf.src_ranges != xf.tgt_ranges:
           if self.version == 1:
diff --git a/tools/releasetools/edify_generator.py b/tools/releasetools/edify_generator.py
old mode 100644
new mode 100755
index 439811c..900fde7
--- a/tools/releasetools/edify_generator.py
+++ b/tools/releasetools/edify_generator.py
@@ -322,7 +322,7 @@ class EdifyGenerator(object):
 
   def SetPermissions(self, fn, uid, gid, mode, selabel, capabilities):
     """Set file ownership and permissions."""
-    if not self.info.get("use_set_metadata", False):
+    if not self.info.get("use_set_metadata", True):
       self.script.append('set_perm(%d, %d, 0%o, "%s");' % (uid, gid, mode, fn))
     else:
       if capabilities is None: capabilities = "0x0"
@@ -335,7 +335,7 @@ class EdifyGenerator(object):
 
   def SetPermissionsRecursive(self, fn, uid, gid, dmode, fmode, selabel, capabilities):
     """Recursively set path ownership and permissions."""
-    if not self.info.get("use_set_metadata", False):
+    if not self.info.get("use_set_metadata", True):
       self.script.append('set_perm_recursive(%d, %d, 0%o, 0%o, "%s");'
                          % (uid, gid, dmode, fmode, fn))
     else:
diff --git a/tools/releasetools/sparse_img.py b/tools/releasetools/sparse_img.py
old mode 100644
new mode 100755
index 7574747..0b9c3c1
--- a/tools/releasetools/sparse_img.py
+++ b/tools/releasetools/sparse_img.py
@@ -41,17 +41,17 @@ class SparseImage(object):
     total_chunks = header[7]
     image_checksum = header[8]
 
-    if magic != 0xED26FF3A:
-      raise ValueError("Magic should be 0xED26FF3A but is 0x%08X" % (magic,))
+#    if magic != 0xED26FF3A:
+#      raise ValueError("Magic should be 0xED26FF3A but is 0x%08X" % (magic,))
     if major_version != 1 or minor_version != 0:
       raise ValueError("I know about version 1.0, but this is version %u.%u" %
                        (major_version, minor_version))
-    if file_hdr_sz != 28:
-      raise ValueError("File header size was expected to be 28, but is %u." %
-                       (file_hdr_sz,))
-    if chunk_hdr_sz != 12:
-      raise ValueError("Chunk header size was expected to be 12, but is %u." %
-                       (chunk_hdr_sz,))
+#    if file_hdr_sz != 28:
+#      raise ValueError("File header size was expected to be 28, but is %u." %
+#                       (file_hdr_sz,))
+#    if chunk_hdr_sz != 12:
+#      raise ValueError("Chunk header size was expected to be 12, but is %u." %
+#                       (chunk_hdr_sz,))
 
     print("Total of %u %u-byte output blocks in %u input chunks."
           % (total_blks, blk_sz, total_chunks))
