diff --git a/Android.mk b/Android.mk
old mode 100644
new mode 100755
diff --git a/Makefile b/Makefile
old mode 100644
new mode 100755
index 1e1dbb3..238c1ab
--- a/Makefile
+++ b/Makefile
@@ -78,7 +78,7 @@ config.h: .config
 
 SUBDIR_VARS := CLEANFILES EXAMPLES FFLIBS HOSTPROGS TESTPROGS TOOLS      \
                HEADERS ARCH_HEADERS BUILT_HEADERS SKIPHEADERS            \
-               ARMV5TE-OBJS ARMV6-OBJS ARMV8-OBJS VFP-OBJS NEON-OBJS     \
+               ARMV5TE-OBJS ARMV6-OBJS ARMV8-OBJS VFP-OBJS     \
                ALTIVEC-OBJS MMX-OBJS YASM-OBJS                           \
                MIPSFPU-OBJS MIPSDSPR2-OBJS MIPSDSPR1-OBJS MIPS32R2-OBJS  \
                OBJS SLIBOBJS HOSTOBJS TESTOBJS
diff --git a/android/config-armv7-a.mak b/android/config-armv7-a.mak
old mode 100644
new mode 100755
index 076e343..fa00133
--- a/android/config-armv7-a.mak
+++ b/android/config-armv7-a.mak
@@ -7,7 +7,7 @@ ARCH_ARM:=yes
 HAVE_ARMV5TE:=yes
 HAVE_ARMV6:=yes
 HAVE_ARMV6T2:=yes
-HAVE_NEON:=yes
+HAVE_NEON:=no
 HAVE_VFP:=yes
 HAVE_VFPV3:=yes
 !HAVE_AMD3DNOW:=yes
diff --git a/android/config.mak b/android/config.mak
old mode 100644
new mode 100755
index 132493d..7e52dfd
--- a/android/config.mak
+++ b/android/config.mak
@@ -769,15 +769,15 @@ CONFIG_VCR1_DECODER=yes
 CONFIG_VMDVIDEO_DECODER=yes
 CONFIG_VMNC_DECODER=yes
 CONFIG_VP3_DECODER=yes
-CONFIG_VP5_DECODER=yes
-CONFIG_VP6_DECODER=yes
-CONFIG_VP6A_DECODER=yes
-CONFIG_VP6F_DECODER=yes
-CONFIG_VP7_DECODER=yes
-CONFIG_VP8_DECODER=yes
-CONFIG_VP9_DECODER=yes
+!CONFIG_VP5_DECODER=yes
+!CONFIG_VP6_DECODER=yes
+!CONFIG_VP6A_DECODER=yes
+!CONFIG_VP6F_DECODER=yes
+!CONFIG_VP7_DECODER=yes
+!CONFIG_VP8_DECODER=yes
+!CONFIG_VP9_DECODER=yes
 CONFIG_VQA_DECODER=yes
-CONFIG_WEBP_DECODER=yes
+!CONFIG_WEBP_DECODER=yes
 CONFIG_WMV1_DECODER=yes
 CONFIG_WMV2_DECODER=yes
 CONFIG_WMV3_DECODER=yes
@@ -1803,8 +1803,8 @@ CONFIG_TAK_PARSER=yes
 CONFIG_VC1_PARSER=yes
 CONFIG_VORBIS_PARSER=yes
 CONFIG_VP3_PARSER=yes
-CONFIG_VP8_PARSER=yes
-CONFIG_VP9_PARSER=yes
+!CONFIG_VP8_PARSER=yes
+!CONFIG_VP9_PARSER=yes
 !CONFIG_BLURAY_PROTOCOL=yes
 CONFIG_CACHE_PROTOCOL=yes
 CONFIG_CONCAT_PROTOCOL=yes
diff --git a/android/include/config-armv7-a.h b/android/include/config-armv7-a.h
old mode 100644
new mode 100755
index a0e3710..36d2c29
--- a/android/include/config-armv7-a.h
+++ b/android/include/config-armv7-a.h
@@ -6,7 +6,7 @@
 #define HAVE_ARMV5TE 1
 #define HAVE_ARMV6 1
 #define HAVE_ARMV6T2 1
-#define HAVE_NEON 1
+#define HAVE_NEON 0
 #define HAVE_VFP 1
 #define HAVE_VFPV3 1
 #define HAVE_AMD3DNOW 0
diff --git a/android/include/config.h b/android/include/config.h
old mode 100644
new mode 100755
index c61b109..3e40c37
--- a/android/include/config.h
+++ b/android/include/config.h
@@ -2,7 +2,7 @@
 #include TARGET_CONFIG
 #ifndef FFMPEG_CONFIG_H
 #define FFMPEG_CONFIG_H
-#define FFMPEG_CONFIGURATION "--arch=arm --target-os=linux --enable-cross-compile --cross-prefix=/a/dev/lollipop/prebuilts/gcc/linux-x86/arm/arm-linux-androideabi-4.8/bin/arm-linux-androideabi- --sysroot=/a/dev/lollipop/prebuilts/ndk/current/platforms/android-19/arch-arm --enable-shared --enable-static --enable-avresample --extra-cflags='-march=armv7-a -mfloat-abi=softfp -mfpu=neon' --extra-ldflags='-Wl,--fix-cortex-a8'"
+#define FFMPEG_CONFIGURATION "--arch=arm --target-os=linux --enable-cross-compile --cross-prefix=/a/dev/lollipop/prebuilts/gcc/linux-x86/arm/arm-linux-androideabi-4.8/bin/arm-linux-androideabi- --sysroot=/a/dev/lollipop/prebuilts/ndk/current/platforms/android-19/arch-arm --enable-shared --enable-static --enable-avresample --extra-cflags='-march=armv7-a -mfloat-abi=softfp' --extra-ldflags='-Wl,--fix-cortex-a8'"
 #define FFMPEG_LICENSE "LGPL version 2.1 or later"
 #define CONFIG_THIS_YEAR 2014
 #define FFMPEG_DATADIR "/usr/local/share/ffmpeg"
@@ -55,7 +55,7 @@
 #define HAVE_ARMV6 1
 #define HAVE_ARMV6T2 1
 #define HAVE_ARMV8 0
-#define HAVE_NEON 1
+#define HAVE_NEON 0
 #define HAVE_VFP 1
 #define HAVE_VFPV3 1
 #define HAVE_SETEND 1
@@ -90,7 +90,7 @@
 #define HAVE_ARMV6_EXTERNAL 1
 #define HAVE_ARMV6T2_EXTERNAL 1
 #define HAVE_ARMV8_EXTERNAL 0
-#define HAVE_NEON_EXTERNAL 1
+#define HAVE_NEON_EXTERNAL 0
 #define HAVE_VFP_EXTERNAL 1
 #define HAVE_VFPV3_EXTERNAL 1
 #define HAVE_SETEND_EXTERNAL 1
@@ -125,7 +125,7 @@
 #define HAVE_ARMV6_INLINE 1
 #define HAVE_ARMV6T2_INLINE 1
 #define HAVE_ARMV8_INLINE 0
-#define HAVE_NEON_INLINE 1
+#define HAVE_NEON_INLINE 0
 #define HAVE_VFP_INLINE 1
 #define HAVE_VFPV3_INLINE 1
 #define HAVE_SETEND_INLINE 1
@@ -224,7 +224,7 @@
 #define HAVE_UNISTD_H 1
 #define HAVE_WINDOWS_H 0
 #define HAVE_WINSOCK2_H 0
-#define HAVE_INTRINSICS_NEON 1
+#define HAVE_INTRINSICS_NEON 0
 #define HAVE_ATANF 1
 #define HAVE_ATAN2F 1
 #define HAVE_CBRT 1
diff --git a/arch.mak b/arch.mak
old mode 100644
new mode 100755
index 0e866d8..882d4a0
--- a/arch.mak
+++ b/arch.mak
@@ -2,7 +2,7 @@ OBJS-$(HAVE_ARMV5TE) += $(ARMV5TE-OBJS) $(ARMV5TE-OBJS-yes)
 OBJS-$(HAVE_ARMV6)   += $(ARMV6-OBJS)   $(ARMV6-OBJS-yes)
 OBJS-$(HAVE_ARMV8)   += $(ARMV8-OBJS)   $(ARMV8-OBJS-yes)
 OBJS-$(HAVE_VFP)     += $(VFP-OBJS)     $(VFP-OBJS-yes)
-OBJS-$(HAVE_NEON)    += $(NEON-OBJS)    $(NEON-OBJS-yes)
+#OBJS-$(HAVE_NEON)    += $(NEON-OBJS)    $(NEON-OBJS-yes)
 
 OBJS-$(HAVE_MIPSFPU)   += $(MIPSFPU-OBJS)    $(MIPSFPU-OBJS-yes)
 OBJS-$(HAVE_MIPS32R2)  += $(MIPS32R2-OBJS)   $(MIPS32R2-OBJS-yes)
diff --git a/libavcodec/Android.mk b/libavcodec/Android.mk
old mode 100644
new mode 100755
index 4adcb5b..6a3b2fd
--- a/libavcodec/Android.mk
+++ b/libavcodec/Android.mk
@@ -21,7 +21,7 @@ LOCAL_SHARED_LIBRARIES +=	\
 	libavutil \
 	libswresample
 
-LOCAL_SRC_FILES += neon/mpegvideo.c
+#LOCAL_SRC_FILES += neon/mpegvideo.c
 
 # It's strange wmalosslessdec.c can't be compiled by -O3 for armv7-a-neon. A gcc bug?
 $(intermediates)/wmalosslessdec.o: PRIVATE_CFLAGS += $(if $(filter arm,$(TARGET_ARCH)),-Os)
@@ -40,7 +40,7 @@ LOCAL_SHARED_LIBRARIES +=	\
 	libavutil \
 	libswresample
 
-LOCAL_SRC_FILES += neon/mpegvideo.c
+#LOCAL_SRC_FILES += neon/mpegvideo.c
 
 # It's strange wmalosslessdec.c can't be compiled by -O3 for armv7-a-neon. A gcc bug?
 $(intermediates)/wmalosslessdec.o: PRIVATE_CFLAGS += $(if $(filter arm,$(TARGET_ARCH)),-Os)
diff --git a/libavcodec/allcodecs.c b/libavcodec/allcodecs.c
old mode 100644
new mode 100755
index 0d39d33..1e24423
--- a/libavcodec/allcodecs.c
+++ b/libavcodec/allcodecs.c
@@ -291,15 +291,15 @@ void avcodec_register_all(void)
     REGISTER_DECODER(VMDVIDEO,          vmdvideo);
     REGISTER_DECODER(VMNC,              vmnc);
     REGISTER_DECODER(VP3,               vp3);
-    REGISTER_DECODER(VP5,               vp5);
-    REGISTER_DECODER(VP6,               vp6);
-    REGISTER_DECODER(VP6A,              vp6a);
-    REGISTER_DECODER(VP6F,              vp6f);
-    REGISTER_DECODER(VP7,               vp7);
-    REGISTER_DECODER(VP8,               vp8);
-    REGISTER_DECODER(VP9,               vp9);
+//    REGISTER_DECODER(VP5,               vp5);
+//    REGISTER_DECODER(VP6,               vp6);
+//    REGISTER_DECODER(VP6A,              vp6a);
+//    REGISTER_DECODER(VP6F,              vp6f);
+//    REGISTER_DECODER(VP7,               vp7);
+//    REGISTER_DECODER(VP8,               vp8);
+//    REGISTER_DECODER(VP9,               vp9);
     REGISTER_DECODER(VQA,               vqa);
-    REGISTER_DECODER(WEBP,              webp);
+//    REGISTER_DECODER(WEBP,              webp);
     REGISTER_ENCDEC (WMV1,              wmv1);
     REGISTER_ENCDEC (WMV2,              wmv2);
     REGISTER_DECODER(WMV3,              wmv3);
@@ -574,8 +574,8 @@ void avcodec_register_all(void)
     REGISTER_PARSER(VC1,                vc1);
     REGISTER_PARSER(VORBIS,             vorbis);
     REGISTER_PARSER(VP3,                vp3);
-    REGISTER_PARSER(VP8,                vp8);
-    REGISTER_PARSER(VP9,                vp9);
+//    REGISTER_PARSER(VP8,                vp8);
+//    REGISTER_PARSER(VP9,                vp9);
 
     /* bitstream filters */
     REGISTER_BSF(AAC_ADTSTOASC,         aac_adtstoasc);
diff --git a/libavcodec/neon/Makefile b/libavcodec/neon/Makefile
old mode 100644
new mode 100755
index 607f116..55d663b
--- a/libavcodec/neon/Makefile
+++ b/libavcodec/neon/Makefile
@@ -1 +1 @@
-OBJS-$(CONFIG_MPEGVIDEO)          += neon/mpegvideo.o
+#OBJS-$(CONFIG_MPEGVIDEO)          += neon/mpegvideo.o
diff --git a/libavutil/arm/Makefile b/libavutil/arm/Makefile
old mode 100644
new mode 100755
index 5da44b0..05eae73
--- a/libavutil/arm/Makefile
+++ b/libavutil/arm/Makefile
@@ -4,5 +4,5 @@ OBJS += arm/cpu.o                                                       \
 VFP-OBJS += arm/float_dsp_init_vfp.o                                    \
             arm/float_dsp_vfp.o                                         \
 
-NEON-OBJS += arm/float_dsp_init_neon.o                                  \
-             arm/float_dsp_neon.o                                       \
+#NEON-OBJS += arm/float_dsp_init_neon.o                                  \
+#             arm/float_dsp_neon.o                                       \
diff --git a/libavutil/arm/float_dsp_init_arm.c b/libavutil/arm/float_dsp_init_arm.c
old mode 100644
new mode 100755
index 6787628..779d9ec
--- a/libavutil/arm/float_dsp_init_arm.c
+++ b/libavutil/arm/float_dsp_init_arm.c
@@ -27,6 +27,6 @@ av_cold void ff_float_dsp_init_arm(AVFloatDSPContext *fdsp)
 
     if (have_vfp(cpu_flags))
         ff_float_dsp_init_vfp(fdsp, cpu_flags);
-    if (have_neon(cpu_flags))
-        ff_float_dsp_init_neon(fdsp);
+//    if (have_neon(cpu_flags))
+//        ff_float_dsp_init_neon(fdsp);
 }
diff --git a/libswresample/arm/Makefile b/libswresample/arm/Makefile
old mode 100644
new mode 100755
index 60f3f6d..2024dd4
--- a/libswresample/arm/Makefile
+++ b/libswresample/arm/Makefile
@@ -1,5 +1,5 @@
 OBJS      += arm/audio_convert_init.o
 
-OBJS-$(CONFIG_NEON_CLOBBER_TEST) += arm/neontest.o
+#OBJS-$(CONFIG_NEON_CLOBBER_TEST) += arm/neontest.o
 
-NEON-OBJS += arm/audio_convert_neon.o
+#NEON-OBJS += arm/audio_convert_neon.o
diff --git a/libswresample/arm/audio_convert_init.c b/libswresample/arm/audio_convert_init.c
old mode 100644
new mode 100755
index ec9e62e..87e9aa9
--- a/libswresample/arm/audio_convert_init.c
+++ b/libswresample/arm/audio_convert_init.c
@@ -31,18 +31,18 @@ void swri_oldapi_conv_fltp_to_s16_2ch_neon(int16_t *dst, float *const *src, int
 void swri_oldapi_conv_fltp_to_s16_nch_neon(int16_t *dst, float *const *src, int len, int channels);
 
 static void conv_flt_to_s16_neon(uint8_t **dst, const uint8_t **src, int len){
-    swri_oldapi_conv_flt_to_s16_neon((int16_t*)*dst, (const float*)*src, len);
+//    swri_oldapi_conv_flt_to_s16_neon((int16_t*)*dst, (const float*)*src, len);
 }
 
 static void conv_fltp_to_s16_2ch_neon(uint8_t **dst, const uint8_t **src, int len){
-    swri_oldapi_conv_fltp_to_s16_2ch_neon((int16_t*)*dst, (float *const*)src, len, 2);
+//    swri_oldapi_conv_fltp_to_s16_2ch_neon((int16_t*)*dst, (float *const*)src, len, 2);
 }
 
 static void conv_fltp_to_s16_nch_neon(uint8_t **dst, const uint8_t **src, int len){
     int channels;
     for(channels=3; channels<SWR_CH_MAX && src[channels]; channels++)
         ;
-    swri_oldapi_conv_fltp_to_s16_nch_neon((int16_t*)*dst, (float *const*)src, len, channels);
+//    swri_oldapi_conv_fltp_to_s16_nch_neon((int16_t*)*dst, (float *const*)src, len, channels);
 }
 
 av_cold void swri_audio_convert_init_arm(struct AudioConvert *ac,
diff --git a/tools/build_libstagefright b/tools/build_libstagefright
old mode 100644
new mode 100755
index 22bb712..ab32395
--- a/tools/build_libstagefright
+++ b/tools/build_libstagefright
@@ -43,7 +43,7 @@ EXTRA_CFLAGS="$EXTRA_CFLAGS -I$ANDROID_SOURCE/frameworks/base/media/libstagefrig
 EXTRA_CFLAGS="$EXTRA_CFLAGS -I$ANDROID_SOURCE/frameworks/base/include/media/stagefright/openmax"
 EXTRA_CFLAGS="$EXTRA_CFLAGS -I$NDK/sources/cxx-stl/gnu-libstdc++/include -I$NDK/sources/cxx-stl/gnu-libstdc++/libs/$ABI/include"
 
-EXTRA_CFLAGS="$EXTRA_CFLAGS -march=armv7-a -mfloat-abi=softfp -mfpu=neon"
+EXTRA_CFLAGS="$EXTRA_CFLAGS -march=armv7-a -mfloat-abi=softfp 
 EXTRA_LDFLAGS="-Wl,--fix-cortex-a8 -L$ANDROID_LIBS -Wl,-rpath-link,$ANDROID_LIBS -L$NDK/sources/cxx-stl/gnu-libstdc++/libs/$ABI"
 EXTRA_CXXFLAGS="-Wno-multichar -fno-exceptions -fno-rtti"
 DEST="$DEST/$ABI"
