diff --git a/src/core/SkBlitMask_D32.cpp b/src/core/SkBlitMask_D32.cpp
old mode 100644
new mode 100755
index fa09b08..e02a961
--- a/src/core/SkBlitMask_D32.cpp
+++ b/src/core/SkBlitMask_D32.cpp
@@ -249,12 +249,11 @@ static void D32_A8_Opaque(void* SK_RESTRICT dst, size_t dstRB,
 #endif
 }
 
-#if !SK_ARM_NEON_IS_NONE
+#if 0
 extern "C" void D32_A8_Black_Neon(void* SK_RESTRICT dst, size_t dstRB,
                          const void* SK_RESTRICT maskPtr, size_t maskRB,
                          SkColor, int width, int height);
 #endif
-
 /*static */void D32_A8_Black(void* SK_RESTRICT dst, size_t dstRB,
                          const void* SK_RESTRICT maskPtr, size_t maskRB,
                          SkColor, int width, int height) {
@@ -582,7 +581,7 @@ static void D32_LCD32_Opaque(void* SK_RESTRICT dst, size_t dstRB,
 
 static SkBlitMask::ColorProc D32_A8_Factory(SkColor color) {
     if (SK_ColorBLACK == color) {
-#if !SK_ARM_NEON_IS_NONE
+#if 0
         return D32_A8_Black_Neon;
 #else
         return D32_A8_Black;
diff --git a/src/core/SkBlitRow_D16.cpp b/src/core/SkBlitRow_D16.cpp
old mode 100644
new mode 100755
index a45d839..da6cfeb
--- a/src/core/SkBlitRow_D16.cpp
+++ b/src/core/SkBlitRow_D16.cpp
@@ -208,7 +208,7 @@ static void S32A_D565_Blend_Dither(uint16_t* SK_RESTRICT dst,
 ///////////////////////////////////////////////////////////////////////////////
 ///////////////////////////////////////////////////////////////////////////////
 
-#if defined(SK_BUILD_FOR_ANDROID) && defined(SK_CPU_LENDIAN)
+#if 0
     extern "C" void S32A_D565_Opaque_arm(uint16_t*, uint32_t*, size_t);
 #endif
 
@@ -217,7 +217,7 @@ static const SkBlitRow::Proc gDefault_565_Procs[] = {
     S32_D565_Opaque,
     S32_D565_Blend,
 
-#if defined(SK_BUILD_FOR_ANDROID) && defined(SK_CPU_LENDIAN)
+#if 0
     (SkBlitRow::Proc)S32A_D565_Opaque_arm,
 #else
     S32A_D565_Opaque,
diff --git a/src/opts/SkBitmapProcState_opts_arm.cpp b/src/opts/SkBitmapProcState_opts_arm.cpp
old mode 100644
new mode 100755
index 8a3f8a4..0f4678b
--- a/src/opts/SkBitmapProcState_opts_arm.cpp
+++ b/src/opts/SkBitmapProcState_opts_arm.cpp
@@ -854,7 +854,9 @@ void SkBitmapProcState::platformProcs() {
                     fSampleProc32 = SI8_opaque_D32_nofilter_DX_arm;
                     fShaderProc32 = NULL;
                 }
+#if defined(__ARM_HAVE_NEON) && defined(SK_CPU_LENDIAN)
             }
+#endif
             break;
         case SkBitmap::kRGB_565_Config:
 #if defined(__ARM_HAVE_NEON) && defined(SK_CPU_LENDIAN)
