diff --git a/crypto/openssl_util.cc b/crypto/openssl_util.cc
index 34af810..23632da
--- a/crypto/openssl_util.cc
+++ b/crypto/openssl_util.cc
@@ -15,7 +15,7 @@
 #include "build/build_config.h"
 
 #if defined(OS_ANDROID) && defined(ARCH_CPU_ARMEL)
-#include <cpu-features.h>
+//#include <cpu-features.h>
 #endif
 
 namespace crypto {
@@ -55,7 +55,7 @@ class OpenSSLInitSingleton {
     CRYPTO_set_locking_callback(LockingCallback);
     CRYPTO_set_id_callback(CurrentThreadId);
 
-#if defined(OS_ANDROID) && defined(ARCH_CPU_ARMEL)
+#if 0
     const bool has_neon =
         (android_getCpuFeatures() & ANDROID_CPU_ARM_FEATURE_NEON) != 0;
     if (has_neon)

