diff --git a/pcm.c b/pcm.c
old mode 100644
new mode 100755
index 394faba..c9c6541
--- a/pcm.c
+++ b/pcm.c
@@ -448,6 +448,8 @@ int pcm_get_htimestamp(struct pcm *pcm, unsigned int *avail,
     int rc;
     snd_pcm_uframes_t hw_ptr;
 
+    if (!pcm || !avail || !tstamp) return -1;
+
     if (!pcm_is_ready(pcm))
         return -1;
 
@@ -455,8 +457,8 @@ int pcm_get_htimestamp(struct pcm *pcm, unsigned int *avail,
     if (rc < 0)
         return -1;
 
-    if ((pcm->mmap_status->state != PCM_STATE_RUNNING) &&
-            (pcm->mmap_status->state != PCM_STATE_DRAINING))
+    if (!pcm || !pcm->mmap_status || !pcm->mmap_control || ((pcm->mmap_status->state != PCM_STATE_RUNNING) &&
+            (pcm->mmap_status->state != PCM_STATE_DRAINING)))
         return -1;
 
     *tstamp = pcm->mmap_status->tstamp;
