diff --git a/pcm.c b/pcm.c
old mode 100644
new mode 100755
index 394faba..d603c04
--- a/pcm.c
+++ b/pcm.c
@@ -455,8 +455,8 @@ int pcm_get_htimestamp(struct pcm *pcm, unsigned int *avail,
     if (rc < 0)
         return -1;
 
-    if ((pcm->mmap_status->state != PCM_STATE_RUNNING) &&
-            (pcm->mmap_status->state != PCM_STATE_DRAINING))
+    if (!pcm || !pcm->mmap_status || !pcm->mmap_control || ((pcm->mmap_status->state != PCM_STATE_RUNNING) &&
+            (pcm->mmap_status->state != PCM_STATE_DRAINING)))
         return -1;
 
     *tstamp = pcm->mmap_status->tstamp;
