diff --git a/policy/src/com/android/internal/policy/impl/PhoneWindowManager.java b/policy/src/com/android/internal/policy/impl/PhoneWindowManager.java
old mode 100644
new mode 100755
index bcfaed9..dfc76fd
--- a/policy/src/com/android/internal/policy/impl/PhoneWindowManager.java
+++ b/policy/src/com/android/internal/policy/impl/PhoneWindowManager.java
@@ -153,7 +153,7 @@ public class PhoneWindowManager implements WindowManagerPolicy {
     static final boolean DEBUG_KEYGUARD = false;
     static final boolean DEBUG_LAYOUT = false;
     static final boolean DEBUG_STARTING_WINDOW = false;
-    static final boolean DEBUG_WAKEUP = false;
+    static final boolean DEBUG_WAKEUP = true;
     static final boolean SHOW_STARTING_ANIMATIONS = true;
     static final boolean SHOW_PROCESSES_ON_ALT_MENU = false;
 
@@ -7101,6 +7101,9 @@ public class PhoneWindowManager implements WindowManagerPolicy {
             case KeyEvent.KEYCODE_HOME:
                 if (DEBUG_WAKEUP) Log.i(TAG, "isOffscreenWakeKey: mHomeWakeSupport " + mHomeWakeSupport);
                 return mHomeWakeSupport;
+            case KeyEvent.KEYCODE_BACK:
+                if (DEBUG_WAKEUP) Log.i(TAG, "isOffscreenWakeKey: mBackWakeSupport 1");
+                return true;
         }
         return false;
     }
