diff --git a/services/core/java/com/android/server/AlarmManagerService.java b/services/core/java/com/android/server/AlarmManagerService.java
old mode 100644
new mode 100755
index 3f59b02..a22b9b6
--- a/services/core/java/com/android/server/AlarmManagerService.java
+++ b/services/core/java/com/android/server/AlarmManagerService.java
@@ -2625,7 +2625,8 @@ class AlarmManagerService extends SystemService {
                                 // is a repeating alarm, so toss the hoser.
                                 removeImpl(alarm.operation);
                             }
-                        }
+                        } catch (Exception e) {
+						}
                     }
                     break;
                 }