diff --git a/media/libstagefright/omx/Android.mk b/media/libstagefright/omx/Android.mk
index aaa8334..e1fc285 100644
--- a/media/libstagefright/omx/Android.mk
+++ b/media/libstagefright/omx/Android.mk
@@ -4,6 +4,9 @@ include $(CLEAR_VARS)
 ifeq ($(TARGET_DEVICE), manta)
     LOCAL_CFLAGS += -DSURFACE_IS_BGR32
 endif
+ifeq ($(BOARD_USES_PROPRIETARY_OMX),TF101)
+LOCAL_CFLAGS     += -DTF101_OMX
+endif
 
 LOCAL_SRC_FILES:=                     \
         GraphicBufferSource.cpp       \
diff --git a/media/libstagefright/omx/OMXMaster.cpp b/media/libstagefright/omx/OMXMaster.cpp
old mode 100644
new mode 100755
index ae3cb33..62352fe
--- a/media/libstagefright/omx/OMXMaster.cpp
+++ b/media/libstagefright/omx/OMXMaster.cpp
@@ -45,6 +45,13 @@ OMXMaster::~OMXMaster() {
 
 void OMXMaster::addVendorPlugin() {
     addPlugin("libstagefrighthw.so");
+#ifdef SAMSUNG_OMX
+    addPlugin("libsomxcore.so");
+#endif
+
+#ifdef TF101_OMX
+    addPlugin("libnvomx.so");
+#endif
 }
 
 void OMXMaster::addPlugin(const char *libname) {
@@ -68,6 +75,9 @@ void OMXMaster::addPlugin(const char *libname) {
 }
 
 void OMXMaster::addPlugin(OMXPluginBase *plugin) {
+    if (plugin == 0) {
+       return;
+    }
     Mutex::Autolock autoLock(mLock);
 
     mPlugins.push_back(plugin);
@@ -106,13 +116,28 @@ void OMXMaster::clearPlugins() {
 
     mPluginByComponentName.clear();
 
+#ifdef TF101_OMX
+    int plugin = 1;
+#endif
     for (List<OMXPluginBase *>::iterator it = mPlugins.begin();
             it != mPlugins.end(); ++it) {
         if (destroyOMXPlugin)
             destroyOMXPlugin(*it);
+#ifdef TF101_OMX
+        else {
+           ALOGV("plugin %d", *it);
+           if( *it != NULL && plugin != 1) // 1st plugin is NV, crashes on delete => skip, better than crash for the moment
+	      delete *it;
+        }
+
+        *it = NULL;
+    plugin++;
+#else
         else
             delete *it;
+
         *it = NULL;
+#endif
     }
 
     mPlugins.clear();
