diff --git a/cmds/installd/commands.cpp b/cmds/installd/commands.cpp
old mode 100644
new mode 100755
index 2a9950a..a56b7fe
--- a/cmds/installd/commands.cpp
+++ b/cmds/installd/commands.cpp
@@ -791,6 +791,14 @@ static void run_dex2oat(int zip_fd, int oat_fd, int image_fd, const char* input_
         sprintf(dex2oat_image_fd, "--app-image-fd=%d", image_fd);
     }
 
+    char dex2oat_compiler_backend[1024];
+    bool have_dex2oat_compiller_backend = false;
+    if (property_get_bool("config.dex2oat-compiler-quick")) {
+        snprintf(dex2oat_compiler_backend, sizeof(dex2oat_compiler_backend),
+            "--compiler-backend=Quick");
+        have_dex2oat_compiller_backend = true;
+    }
+
     if (have_dex2oat_Xms_flag) {
         sprintf(dex2oat_Xms_arg, "-Xms%s", dex2oat_Xms_flag);
     }
@@ -849,6 +857,7 @@ static void run_dex2oat(int zip_fd, int oat_fd, int image_fd, const char* input_
                      + (have_dex2oat_swap_fd ? 1 : 0)
                      + (have_dex2oat_image_fd ? 1 : 0)
                      + (have_dex2oat_relocation_skip_flag ? 2 : 0)
+                     + (have_dex2oat_compiller_backend ? 1 : 0)
                      + (generate_debug_info ? 1 : 0)
                      + (debuggable ? 1 : 0)
                      + (have_app_image_format ? 1 : 0)
@@ -888,6 +897,9 @@ static void run_dex2oat(int zip_fd, int oat_fd, int image_fd, const char* input_
     if (have_dex2oat_image_fd) {
         argv[i++] = dex2oat_image_fd;
     }
+    if (have_dex2oat_compiller_backend) {
+        argv[i++] = dex2oat_compiler_backend;
+    }
     if (generate_debug_info) {
         argv[i++] = "--generate-debug-info";
     }
