diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
old mode 100644
new mode 100755
index 92fe981..f2ab3aa
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -1411,7 +1411,7 @@ public final class Settings {
          * stored as strings, so this function converts the given value to a
          * string before storing it.
          *
-         * @param cr The ContentResolver to access.
+         * @param resolver The ContentResolver to access.
          * @param name The name of the setting to modify.
          * @param value The new value for the setting.
          * @return true if the value was set, false on database errors
diff --git a/core/java/android/widget/SeekBar.java b/core/java/android/widget/SeekBar.java
old mode 100644
new mode 100755
index dc7c04c..0545035
--- a/core/java/android/widget/SeekBar.java
+++ b/core/java/android/widget/SeekBar.java
@@ -49,7 +49,7 @@ public class SeekBar extends AbsSeekBar {
          * 
          * @param seekBar The SeekBar whose progress has changed
          * @param progress The current progress level. This will be in the range 0..max where max
-         *        was set by {@link ProgressBar#setMax(int)}. (The default value for max is 100.)
+         *        was set by ProgressBar#setMax(int). (The default value for max is 100.)
          * @param fromUser True if the progress change was initiated by the user.
          */
         void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser);
diff --git a/libs/hwui/tests/Android.mk b/libs/hwui/tests/Android.mk
old mode 100644
new mode 100755
index 3d405ef..7bdce7f
--- a/libs/hwui/tests/Android.mk
+++ b/libs/hwui/tests/Android.mk
@@ -50,6 +50,7 @@ LOCAL_MULTILIB := both
 LOCAL_MODULE_STEM_32 := hwuitest
 LOCAL_MODULE_STEM_64 := hwuitest64
 
+include external/stlport/libstlport.mk
 include $(BUILD_EXECUTABLE)
 
 include $(call all-makefiles-under,$(LOCAL_PATH))
