*** cm10.2/dalvik/vm/Init.cpp	2013-12-23 18:14:06.468469000 -0700
--- cm10.2mods/dalvik/vm/Init.cpp	2013-12-23 18:16:56.020463100 -0700
***************
*** 1696,1702 ****
      const char* target_base = getenv("EMULATED_STORAGE_TARGET");
      if (target_base != NULL) {
          if (mount("tmpfs", target_base, "tmpfs", MS_NOSUID | MS_NODEV,
!                 "uid=0,gid=1028,mode=0050") == -1) {
              SLOGE("Failed to mount tmpfs to %s: %s", target_base, strerror(errno));
              return -1;
          }
--- 1696,1702 ----
      const char* target_base = getenv("EMULATED_STORAGE_TARGET");
      if (target_base != NULL) {
          if (mount("tmpfs", target_base, "tmpfs", MS_NOSUID | MS_NODEV,
!                 "uid=0,gid=1028,mode=0050,fscontext=u:object_r:sdcard:s0") == -1) {
              SLOGE("Failed to mount tmpfs to %s: %s", target_base, strerror(errno));
              return -1;
          }
