*** a/frameworks/base/core/java/android/inputmethodservice/IInputMethodWrapper.java	2014-03-30 10:29:00.301005939 -0600
--- b/frameworks/base/core/java/android/inputmethodservice/IInputMethodWrapper.java	2014-03-30 10:29:04.362654346 -0600
***************
*** 284,289 ****
--- 284,293 ----
      @Override
      public void setSessionEnabled(IInputMethodSession session, boolean enabled) {
          try {
+ 			if (session == null) {
+                 Log.w(TAG, "No Session");
+                 return;
+ 			}
              InputMethodSession ls = ((IInputMethodSessionWrapper)
                      session).getInternalInputMethodSession();
              if (ls == null) {
