*** /dev/null	2014-03-04 11:01:47.868460615 -0700
--- vendor/pac/products/pac_adam_3g.mk	2014-02-17 17:46:46.169169200 -0700
***************
*** 0 ****
--- 1,29 ----
+ # Check for target product
+ ifeq (pac_adam_3g,$(TARGET_PRODUCT))
+ 
+ # OVERLAY_TARGET adds overlay asset source
+ OVERLAY_TARGET := adam_3g
+ 
+ # AOKP device overlay
+ PRODUCT_PACKAGE_OVERLAYS += vendor/pac/overlay/aokp/common_tablet
+ 
+ # PAC device overlay
+ PRODUCT_PACKAGE_OVERLAYS += vendor/pac/overlay/pac/hdpi
+ 
+ # PAC boot logo
+ PRODUCT_COPY_FILES += \
+     vendor/pac/prebuilt/common/bootlogo/pac_logo_1024x600.rle:root/logo.rle
+ 
+ # Copy bootanimation
+ PRODUCT_COPY_FILES += \
+     vendor/pac/prebuilt/1280x800/bootanimation.zip:system/media/bootanimation.zip
+ 
+ # include PAC common configuration
+ include vendor/pac/config/pac_common.mk
+ 
+ # Inherit CM device configuration
+ $(call inherit-product, device/notionink/adam_3g/cm.mk)
+ 
+ PRODUCT_NAME := pac_adam_3g
+ 
+ endif
