*** omnirom4.4/frameworks/base/services/java/com/android/server/power/PowerManagerService.java	2014-02-06 19:10:42.029189900 -0700
--- omnirom4.4mods/frameworks/base/services/java/com/android/server/power/PowerManagerService.java	2014-02-10 19:44:32.230448600 -0700
***************
*** 798,809 ****
              throw new IllegalArgumentException("lock must not be null");
          }
  
!         mContext.enforceCallingOrSelfPermission(android.Manifest.permission.WAKE_LOCK, null);
!         if (ws != null && ws.size() != 0) {
!             mContext.enforceCallingOrSelfPermission(
                      android.Manifest.permission.UPDATE_DEVICE_STATS, null);
!         } else {
!             ws = null;
          }
  
          final long ident = Binder.clearCallingIdentity();
--- 798,813 ----
              throw new IllegalArgumentException("lock must not be null");
          }
  
!         
!         
!         if (mContext != null) {
!            mContext.enforceCallingOrSelfPermission(android.Manifest.permission.WAKE_LOCK, null);
!            if (ws != null && ws.size() != 0) {
!                 mContext.enforceCallingOrSelfPermission(
                      android.Manifest.permission.UPDATE_DEVICE_STATS, null);
!            } else {
!                 ws = null;
!            }
          }
  
          final long ident = Binder.clearCallingIdentity();
