diff --git a/bta/dm/bta_dm_act.c b/bta/dm/bta_dm_act.c
old mode 100644
new mode 100755
diff --git a/bta/dm/bta_dm_api.c b/bta/dm/bta_dm_api.c
old mode 100644
new mode 100755
diff --git a/bta/dm/bta_dm_main.c b/bta/dm/bta_dm_main.c
old mode 100644
new mode 100755
diff --git a/btif/include/btif_common.h b/btif/include/btif_common.h
old mode 100644
new mode 100755
index 7e8c506..8f7892c
--- a/btif/include/btif_common.h
+++ b/btif/include/btif_common.h
@@ -27,6 +27,8 @@
 #include "bta_api.h"
 #include "osi.h"
 
+#undef TEST_APP_INTERFACE
+
 #ifndef LOG_TAG
 #error "LOG_TAG not defined, please add in .c file prior to including bt_common.h"
 #endif
diff --git a/btif/include/btif_gatt_multi_adv_util.h b/btif/include/btif_gatt_multi_adv_util.h
old mode 100644
new mode 100755
diff --git a/btif/src/bluetooth.c b/btif/src/bluetooth.c
old mode 100644
new mode 100755
diff --git a/btif/src/btif_core.c b/btif/src/btif_core.c
old mode 100644
new mode 100755
diff --git a/btif/src/btif_dm.c b/btif/src/btif_dm.c
old mode 100644
new mode 100755
diff --git a/hci/src/vendor.c b/hci/src/vendor.c
old mode 100644
new mode 100755
diff --git a/main/bte_logmsg.c b/main/bte_logmsg.c
old mode 100644
new mode 100755
diff --git a/stack/btm/btm_acl.c b/stack/btm/btm_acl.c
old mode 100644
new mode 100755
diff --git a/stack/btm/btm_int.h b/stack/btm/btm_int.h
old mode 100644
new mode 100755
