diff --git a/adb/Android.mk b/adb/Android.mk
old mode 100644
new mode 100755
index e841205..88f2454
--- a/adb/Android.mk
+++ b/adb/Android.mk
@@ -112,6 +112,7 @@ LOCAL_SANITIZE := $(adb_target_sanitize)
 # Even though we're building a static library (and thus there's no link step for
 # this to take effect), this adds the includes to our path.
 LOCAL_STATIC_LIBRARIES := libcrypto_utils libcrypto libbase
+LOCAL_C_INCLUDES := system/core/base/include
 
 include $(BUILD_STATIC_LIBRARY)
 
@@ -131,6 +132,7 @@ LOCAL_SANITIZE := $(adb_target_sanitize)
 LOCAL_STATIC_LIBRARIES := libcrypto_utils libcrypto libbase
 
 LOCAL_WHOLE_STATIC_LIBRARIES := libadbd_usb
+LOCAL_C_INCLUDES := system/core/base/include
 
 include $(BUILD_STATIC_LIBRARY)
 
@@ -158,6 +160,8 @@ LOCAL_STATIC_LIBRARIES := libcrypto_utils libcrypto libbase libmdnssd
 LOCAL_STATIC_LIBRARIES_linux := libusb
 LOCAL_STATIC_LIBRARIES_darwin := libusb
 
+LOCAL_C_INCLUDES := system/core/base/include \
+	external/boringssl/include
 LOCAL_C_INCLUDES_windows := development/host/windows/usb/api/
 LOCAL_MULTILIB := first
 
@@ -178,6 +182,7 @@ LOCAL_SRC_FILES := \
 LOCAL_SANITIZE := $(adb_target_sanitize)
 LOCAL_STATIC_LIBRARIES := libadbd libcrypto_utils libcrypto libusb libmdnssd
 LOCAL_SHARED_LIBRARIES := liblog libbase libcutils
+LOCAL_C_INCLUDES := system/core/base/include
 include $(BUILD_NATIVE_TEST)
 
 # libdiagnose_usb
@@ -191,6 +196,7 @@ LOCAL_SRC_FILES := diagnose_usb.cpp
 # Even though we're building a static library (and thus there's no link step for
 # this to take effect), this adds the includes to our path.
 LOCAL_STATIC_LIBRARIES := libbase
+LOCAL_C_INCLUDES := system/core/base/include
 include $(BUILD_HOST_STATIC_LIBRARY)
 
 # adb_test
@@ -238,6 +244,7 @@ LOCAL_LDLIBS_windows := -lws2_32 -luserenv
 LOCAL_STATIC_LIBRARIES_windows := AdbWinApi
 
 LOCAL_MULTILIB := first
+LOCAL_C_INCLUDES := system/core/base/include
 
 include $(BUILD_HOST_NATIVE_TEST)
 
@@ -308,7 +315,7 @@ LOCAL_CXX_STL := libc++_static
 # on the host adb tool, and shared libraries that link against libc++
 # will violate ODR
 LOCAL_SHARED_LIBRARIES :=
-
+LOCAL_C_INCLUDES := system/core/base/include
 include $(BUILD_HOST_EXECUTABLE)
 
 $(call dist-for-goals,dist_files sdk win_sdk,$(LOCAL_BUILT_MODULE))
@@ -377,6 +384,8 @@ LOCAL_STATIC_LIBRARIES := \
     libmdnssd \
     libdebuggerd_handler \
 
+LOCAL_C_INCLUDES := system/core/base/include
+
 include $(BUILD_EXECUTABLE)
 
 include $(call first-makefiles-under,$(LOCAL_PATH))
