*** /dev/null	2014-03-04 11:01:47.868460615 -0700
--- b/products/pac_adam_3g.mk	2014-02-17 17:46:46.169169200 -0700
***************
*** 0 ****
--- 1,18 ----
+ # Check for target product
+ ifeq (pac_adam_3g,$(TARGET_PRODUCT))
+ 
+ # OVERLAY_TARGET adds overlay asset source
+ OVERLAY_TARGET := adam_3g
+ 
+ # Bootanimation
+ PAC_BOOTANIMATION_NAME := 480
+ 
+ # include PAC common configuration
+ include vendor/pac/config/pac_common.mk
+ 
+ # Inherit CM device configuration
+ $(call inherit-product, device/notionink/adam_3g/cm.mk)
+ 
+ PRODUCT_NAME := pac_adam_3g
+ 
+ endif
