diff --git a/config/common.mk b/config/common.mk
old mode 100644
new mode 100755
index 250819a..1b03e7d
--- a/config/common.mk
+++ b/config/common.mk
@@ -41,8 +41,8 @@ PRODUCT_COPY_FILES += \
     vendor/slim/prebuilt/common/etc/init.local.rc:root/init.slim.rc
 
 # Copy latinime for gesture typing
-PRODUCT_COPY_FILES += \
-    vendor/slim/prebuilt/common/lib/libjni_latinimegoogle.so:system/lib/libjni_latinimegoogle.so
+#PRODUCT_COPY_FILES += \
+#    vendor/slim/prebuilt/common/lib/libjni_latinimegoogle.so:system/lib/libjni_latinimegoogle.so
 
 # SELinux filesystem labels
 PRODUCT_COPY_FILES += \
diff --git a/sepolicy/userinit.te b/sepolicy/userinit.te
old mode 100644
new mode 100755
index 7407287..cb8472b
--- a/sepolicy/userinit.te
+++ b/sepolicy/userinit.te
@@ -1,4 +1,3 @@
-type userinit_exec, exec_type, file_type;
-type userinit_data_exec, file_type;
+type userinit_data_exec;
 
 allow userinit_exec userinit_prop:property_service set;
